libs/wingc/src/type_check/fqn.rs

```diff
@@ -0,0 +1,161 @@
+use std::fmt::{Display, Formatter};
+
+use crate::{CONSTRUCT_BASE, WINGSDK_ASSEMBLY_NAME, WINGSDK_RESOURCE};
+
+/// Represents a fully-qualified name (FQN) of a type in a JSII library.
+/// For example, `@aws-cdk/aws-ec2.Vpc` is a FQN.
+/// The FQN uniquely identifies a type.
+///
+/// The FQN typically looks like:
+/// `assembly_name.namespace1.namespace2...namespaceN.type_name`
+/// where there may be zero or more namespaces.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct FQN<'a>(&'a str);
+
+impl<'a> From<&'a str> for FQN<'a> {
+	fn from(fqn: &'a str) -> Self {
+		if fqn.split('.').count() < 2 {
+			panic!("Invalid FQN: {}", fqn);
+		}
+		FQN(fqn)
+	}
+}
+
+impl<'a> From<&'a String> for FQN<'a> {
```

Comment:
Not sure, I got hit several obstacles trying to de-duplicate these. This feels like it could be a Rust learning opportunity, so I've opened a StackOverflow question: https://stackoverflow.com/questions/75378250/how-to-reuse-implementation-of-fromstr-and-fromstring-in-rust

Reply:
Ok, yeah you were right - I've just removed it and called `.as_str()` to convert where needed